### DISCUSSION TITLE
Intermittent StaleDataError exceptions with minimal table usage and spontaneous resolution

### DISCUSSION THREAD

**Discussion by neuron-whisperer** [sqlalchemy/sqlalchemy#8834] [Usage Questions]
### Describe the bug

I'm using SQLAlchemy to interface with a small Google Cloud SQL database using the imperative mode. The database is extremely vanilla - about 40 tables, autoincrementing integer primary keys on all tables, and a few single-attribute foreign key relationships.

During a data load operation, while updating rows / objects in a table with about 1,300 records, my script threw this exception on record 47:

> sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'applications' expected to update 1 row(s); 0 were matched.

Very puzzling since the same operation was being performed on the 47th record as on the previous 46.

I ran the script again, repeatedly, and got the same exception *on the very first row*.

I performed some regression testing, taking out anything that might have modified the object, all the way down to this:

    applications = list(orm.retrieve_objects('applications'))    # retrieve objects from applications table
    for a in applications:
        a.some_date_field = datetime.date.today()
        orm.commit()

...and the script kept throwing this exception, on the first row, repeatedly. Commenting out the field update caused the script to run fine (as it wasn't updating any records); uncommenting the field update caused the exception to recur. I observed this repeatedly.

Finally, I tweaked the code like this:

    applications = list(orm.retrieve_objects('applications'))    # retrieve objects from applications table
    for a in applications:
        a.some_date_field = a.some_date_field
        orm.commit()

...and the exception disappeared. Then, when I reverted to the original code, the exception didn't recur. The code now runs fine. (I would love to provide SQL log output as suggested in the GitHub bug-report guidelines, but obviously I can't, given the spontaneous resolution.)

Based on this experience, I have serious concerns about the prospect of an intermittent StaleDataError that manifests without explanation and spontaneously resolves.

### To Reproduce

```python
This code intermittently provoked StaleDataErrors:

    applications = list(orm.retrieve_objects('applications'))    # retrieve objects from applications table
    for a in applications:
        a.some_date_field = datetime.date.today()
        orm.commit()
```

### Error

Processing application 44 / 1306 (3.4%)
Processing application 45 / 1306 (3.4%)
Processing application 46 / 1306 (3.5%)
Processing application 47 / 1306 (3.6%)
Traceback (most recent call last):
...
raise orm_exc.StaleDataError(
sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'applications' expected to update 1 row(s); 0 were matched.

### Versions

- OS: MacOS 12.4
- Python: 3.10.4
- SQLAlchemy: 1.4.4
- Database: Google Cloud SQL
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): Google Cloud SQL Python Connector, version 0.9.9

**Comment by zzzeek** [comment]
this error means the ORM is attempting to UPDATE a row by primary key, in this case we'd assume a single row (Are you flushing one record at a time?), and when the UPDATE runs, it delivers that there were no rows affected, when it was expecting one.

Reasons for this include:

1. the row that you loaded, then updated, was deleted either by some other part of your program by the time the Session flushed, or by another server process.
2. or, same thing as 1 but vastly less common unless your program UPDATEs primary key values, the primary key was modified for the target row by another part of the program / other process, leading it to no longer be present.
3. the behavior of `cursor.rowcount` is not consistent on your target driver.   more common than one would think, though it does not tend to fail on random rows.   However, some kinds of clustered databases with various forms of primary/replica and/or eventual consistency might have strange behaviors here.   

the thing you want to check when this UPDATE fails is: "SELECT * FROM table WHERE primary_key_column=<the primary key>".  That's the row it's looking for.  if it's not there, you will get this error.

**Reply by neuron-whisperer** [reply]
I understand all of that. All three of the reasons you provided above make sense.

But none of that applies to my case.

Again - only I am using this database, and I am only accessing it via this one script, one instance at a time. And the code above is all that's happening - basically:

1) Load objects from table

2) Iterate over each object

3) For each object, change some fields (not including the id / key) and then orm.commit() after updating each object

(Why am I updating one object at a time? Because processing each object takes a while - about 30 seconds each, which involves about a 29.9-second delay. There's no reason not to commit it immediately.)

The code couldn't be simpler. So unless something is going wrong with Google Cloud SQL - where the database is spontaneously changing somehow between selecting the objects and re-writing them - then there's a bug with either SQLAlchemy or the Google API Connector, where the version of the row that it reads does not match the version of the row that's later updated.

I usually work through these issues by just trying things and banging my head against the problem until a solution appears, but the simplicity of this scenario makes the problem both baffling and troubling - hence my post here today.

**Reply by zzzeek** [reply]
So first off I'd note that there are sometimes extremely simple bits of code that have some odd thing, which is very hard to spot, that can cause this kind of problem.  so while you can likely rule out item 2, until you identify the issue, item 1 should not be discounted.   

however for a remote, opaque product like google cloud SQL, item 3 is very much in play.

The main thing that definitely applies from the answer I gave is to add debugging that will do a SELECT from the table in question with the failing primary key value, ensuring that the row is in fact present.  you also want to [turn on debug logging](https://docs.sqlalchemy.org/en/14/core/engines.html#configuring-logging) so that all the queries emitted, as well as rows returned, are logged.  you would be able to see a row coming back with a particular primary key, then see the UPDATE statement using that particular primary key, all of which would indicate, in the absence of other statements in between that also affect that same key, that something is up with the driver.

Another thing to point out, since I have no access to your models or queries or anything like that, there can be datatype issues for this kind of thing also.   I notice you're updating a datetime field.  if for example the primary key of these objects used a datetime as part of the key, issues like microseconds not being returned in the value, which then causes the UPDATE to not match, or things like integer identifiers that for whatever reason are formatted as strings on the python side  (in the logs there's a difference between a parameter that looks like ``3`` vs. one that looks like ``'3'``, for example), these are all things that you need to look for.  the SQL echo output is the first step to getting visibility on this.

**Comment by neuron-whisperer** [comment]
I understand your point, but I just encountered the problem again and managed to capture the SQL that provoked the error. And the log seems to suggest that something truly weird is going on.

I'm using some caveman debugging to isolate the location. Going back to these two lines of code, I have this:

    print('C')
    a.some_date_field = a.some_date_field
    orm.commit()

...and here's the console output leading up to the exception (adding line breaks for readabillity):

    C

    INFO:sqlalchemy.engine.Engine:SELECT applications.id AS applications_id,
    (...big list of other attributes to retrieve, followed by...)
    WHERE applications.id = %(pk_1)s

    INFO:sqlalchemy.engine.Engine:[generated in 0.00031s] {'pk_1': 2}

    INFO:sqlalchemy.engine.Engine:UPDATE applications SET prosecution_update=
    %(prosecution_update)s WHERE applications.id = %(applications_id)s

    INFO:sqlalchemy.engine.Engine:[generated in 0.00028s] {'prosecution_update':
    datetime.date(2022, 11, 18), 'applications_id': 2}

    INFO:sqlalchemy.engine.Engine:ROLLBACK

    (stack trace)

    sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'applications'
    expected to update 1 row(s); 0 were matched.

Here's how I interpret the output:

* INFO statement 1: The first SQL statement requests all attributes for the object with primary key (i.e., id) = 2.

* INFO statement 2: Found a row with applications.id (primary key) = 2.

* INFO statements 3 and 4: Update the row where applications.id = a particular value, where the particular value is 2.

* INFO statement 5: ROLLBACK because something failed.

* StaleDataError: Last SQL statement (requesting row with pk = 2) failed.

This appears to confirm what I believe is happening: SQLAlchemy successfully retrieves a row (corresponding to an object) with a particular primary key value, and then immediately fails to update another attribute of the row in that table that has the exact same primary key value.

Do you interpret this differently?

**Reply by zzzeek** [reply]
that log output is all INFO level, rows would be indicated on DEBUG level lines.  those dictionary outputs are the parameters passed to the statement, not rows.  to see rows returned, the level has to be set to DEBUG.

if you want to turn off any reliance on "rowcount", you can set `engine.dialect.supports_sane_rowcount = engine.dialect.supports_sane_multi_rowcount = False`.

But in that scenario, you won't get any errors.  you can try using SELECT after the fact to confirm that everything is being UPDATEd as expected but that won't tell us if the rowcount function itself is where the problem is.

While we are on those attributes, I would note they indicate two different things.  one is if cursor.rowcount works after a cursor.execute() and the other if cursor.rowcount works after a cursor.executemany().  The latter is often not supported by DBAPIs.  if the driver you are using has not set this attribute correctly, that could cause sporadic issues; the ORM uses execute() or executemany() within a flush based on if many UPDATE statements at once have the identical structure.  Though it wouldn't be just "1 row matched" in that case.

**Reply by zzzeek** [reply]
hey good morning!  I just thought of the other thing that can be the problem here, it's something I havent ever seen outside of MySQL but if this database is considering "rowcount" to be "number of rows **affected**" instead of "number of rows **matched**", then you could certainly see this problem exactly as it's manifesting; an UPDATE statement that changes values in the row returns it as a "match", but then once you run that statement repeatedly with the same data, there is no net change and it returns no match.

In MySQL this is documented under CLIENT_FOUND_ROWS at https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_row-count, we set this parameter for the MySQL dialect.

Google cloud SQL would need a way for it to return to us all "matched" rows and not just "affected" rows, otherwise "supports_sane_rowcount" needs to be false.  This would be a bug in the dialect you're using.
